### PR TITLE
access log / route headers: Expose more downstream TLS information

### PR DESCRIPTION
**Pull request by mikegrass**
Description:
* New accessors added for querying information about the established TLS session: issuerPeerCertificate(), ciphersuiteId(), ciphersuiteString(), tlsVersion().
* access_log_formatter.cc and header_formatter.cc updated with new variables for exposing downstream TLS information.
* grpc_access_log_impl.cc updated to fill in existing (but previously-hidden) ciphersuite and tls version fields.
* Unit tests updated to exercise the above changes.

Risk Level: Low
Testing: Unit tests and ad-hoc manual testing
Docs Changes: Updated [docs/root/configuration/access_log.rst](https://github.com/envoyproxy/envoy/compare/master...mikegrass:tls-headers?expand=1#diff-57186174d5943eb3a6b9122dea30be94) and [docs/root/configuration/http_conn_man/headers.rst](https://github.com/envoyproxy/envoy/compare/master...mikegrass:tls-headers?expand=1#diff-db458208464b55cb68c1f038faf7dc23) with information on new variables.
Release Notes: Updated [docs/root/intro/version_history.rst](https://github.com/envoyproxy/envoy/compare/master...mikegrass:tls-headers?expand=1#diff-d8d3e33358b55e6c0466dd1bb5f40c79) to note the changes.

More incremental progress on #3296.

### CODE REVIEW

**Comment by mikegrass** [comment]
@mattklein123 thanks for assigning a reviewer! @zuercher looking forward to your feedback!

**Comment by mikegrass** [comment]
@zuercher -- anything I can do to help move this review along?

File: docs/root/configuration/access_log.rst [hunk 1]
```diff
@@ -360,9 +360,57 @@ The following command operators are supported:
   TCP
     The subject present in the peer certificate used to establish the downstream TLS connection.
 
+%DOWNSTREAM_PEER_ISSUER%
```
Comment (zuercher) [inline docs/root/configuration/access_log.rst]: Should this be DOWNSTREAM_TLS_PEER_ISSUER to make it clear this refers to TLS.

File: source/extensions/transport_sockets/tls/utility.cc [hunk 2]
```diff
@@ -10,6 +10,46 @@ namespace Extensions {
...
+
+  X509_NAME* name = nullptr;
+  switch (desired_name) {
+  case CertName::Issuer:
+    name = X509_get_issuer_name(&cert);
+    break;
+  case CertName::Subject:
+    name = X509_get_subject_name(&cert);
+    break;
+  }
```
Comment (zuercher) [inline source/extensions/transport_sockets/tls/utility.cc]: Usually we'll add a default case with a `NOT_REACHED_GCOVR_EXCL_LINE` statement to help detect missing cases.

File: docs/root/configuration/access_log.rst (hunk 1 above)
Comment (mikegrass) [reply docs/root/configuration/access_log.rst]: Good question, and one that I wrestled with as well. In a clean room, I would have preferred `DOWNSTREAM_TLS_PEER_ISSUER`; however, there was already some precedent set in https://github.com/envoyproxy/envoy/commit/65c05cad96db7f7874e1e5a2b70829dcfad10e88 (included in 1.10) for omitting the `TLS` prefix when referring to the peer and local certificates in access logs.

I opted for consistency with that precedent, but perhaps I should revisit? Let me know what you think.

File: source/extensions/transport_sockets/tls/utility.cc (hunk 2 above)
Comment (mikegrass) [reply source/extensions/transport_sockets/tls/utility.cc]: Fixed, thank you!

File: docs/root/configuration/access_log.rst (hunk 1 above)
Comment (zuercher) [reply docs/root/configuration/access_log.rst]: Ah, yeah. I guess consistency is probably best.

**Comment by mikegrass** [comment]
Thanks for the review @zuercher! Is there anyone else who needs to review before this can be merged?

File: source/extensions/transport_sockets/tls/utility.cc (hunk 2 above)
Comment (mattklein123) [reply source/extensions/transport_sockets/tls/utility.cc]: This default shouldn't be needed at all since you cover all cases. Can you remove?

Comment (mikegrass) [reply source/extensions/transport_sockets/tls/utility.cc]: @mattklein123 Sure -- I've removed it in the latest commit.

**Comment by mattklein123** [comment]
@mikegrass friendly request to not force push in the future, it makes reviews more difficult. Thank you!

**Comment by mikegrass** [comment]
> @mikegrass friendly request to not force push in the future, it makes reviews more difficult. Thank you!

@mattklein123 apologies for that! I had moved to a new computer just yesterday and had neglected to run the `support/bootstrap` script to set up git hooks, so my last commit was missing DCO. I followed the instructions at https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco to fix DCO on that last commit.

Thank you for the review!

**Pull request merged**